stac_generator/base/generator.py

```diff
@@ -1,83 +1,113 @@
-        kwargs = {"source_type": source_type}
-        return type(f"StacGenerator[{source_type.__name__}]", (StacGenerator,), kwargs)
+logger = logging.getLogger(__name__)
+
+
+class CollectionGenerator:
+    """CollectionGenerator class. User should not need to subclass this class unless greater control over how collection is generated from items is needed."""
 
     def __init__(
         self,
-        source_df: pd.DataFrame,
         collection_cfg: StacCollectionConfig,
-        catalog_cfg: StacCatalogConfig | None = None,
-        href: str | None = None,
+        generators: Sequence[ItemGenerator],
     ) -> None:
-        """Base STACGenerator object. Users should extend this class for handling different file extensions.
-        Please see CsvStacGenerator source code.
+        """CollectionGenerator - generate collection from generators attribute
 
-        :param source_df: source data config
-        :type source_df: pd.DataFrame
-        :param collection_cfg: collection config for generating collection. Required if the intention is to save the catalog to disk or POST to an API, defaults to None
+        :param collection_cfg: collection metadata
         :type collection_cfg: StacCollectionConfig
-        :param catalog_cfg: catalog config for generating catalog. Required if the intention is to save the catalog to disk or POST to an API, defaults to None
-        :type catalog_cfg: StacCatalogConfig | None, optional
-        :param href: catalog href for serialisation.
-        :type href: str | None, optional
-
+        :param generators: sequence of ItemGenerator subclasses.
+        :type generators: Sequence[ItemGenerator]
         """
         self.collection_cfg = collection_cfg
-        self.catalog_cfg = (
-            catalog_cfg
-            if catalog_cfg
-            else StacCatalogConfig(
-                id=collection_cfg.id,
-                title=collection_cfg.title,
-                description=collection_cfg.description,
-            )
-        )
-        self.source_df = source_df
-        self.configs = [
-            self.source_type(**cast(dict[str, Any], self.source_df.loc[i, :].to_dict()))
-            for i in range(len(self.source_df))
-        ]
-        self.href = href
+        self.generators = generators
 
-    @abc.abstractmethod
-    def create_item_from_config(self, source_cfg: T) -> list[pystac.Item]:
-        """Abstract method that handles `pystac.Item` generation from the appropriate config"""
-        raise NotImplementedError
+    @staticmethod
+    def spatial_extent(items: Sequence[pystac.Item]) -> pystac.SpatialExtent:
+        """Extract a collection's spatial extent based on geometry information of its items
+
+        :param items: a list of children items in the collection
+        :type items: Sequence[pystac.Item]
+        :return: a bbox enveloping all items
+        :rtype: pystac.SpatialExtent
+        """
+        geometries: list[Geometry] = []
```

Comment:
This is an overcomplication. I think the simplest method is to find the bbox of the union of the bboxes of items. You do not to make a new geodataframe to do this. There are in-built shapely union functions .

Reply:
Fixed

Reply:
Actually got a very weird bug. I replace the method with the official API: 

```python 
import shapely 

@staticmethod
    def spatial_extent(items: Sequence[pystac.Item]) -> pystac.SpatialExtent:
        bboxes = [box(*item.bbox) for item in items]
        bbox = list(shapely.ops.unary_union(bboxes).bounds)
        return pystac.SpatialExtent(bbox)
```
This is the bbox for all items: 

```
[[138.5196, -34.9524, 138.5196, -34.9524], [138.6434, -34.7674, 138.6434, -34.7674], [138.727, -34.9423, 138.727, -34.9423], [138.6498, -35.0076, 138.6498, -35.0076], [138.6223, -34.7111, 138.6223, -34.7111], [138.6519, -34.9464, 138.6519, -34.9464], [138.5662, -35.0591, 138.5662, -35.0591], [138.6844, -34.8564, 138.6844, -34.8564], [138.595, -34.9163, 138.595, -34.9163], [150.5471916, -24.34091973, 150.5560285, -24.33600118]]
```

Note how minimum longitude value is -138. something. The enclosing bbox is 

```
[150.5471916, -24.34091973, 150.5560285, -24.33600118]
```
which is different from the actual

```
[[138.5196, -35.0591, 150.5560285, -24.33600118]]
``` 
Anyway I just realise that both implementations are incorrect since they don't take into account the CRS. Will need a different fix.

Reply:
For now we can assume all items in a collection have the same CRS and solve simply. Handle the more difficult case in the future.